### PR TITLE
The Lambda onboarder

### PR DESCRIPTION
**Pull request by AlexeyBeley**
Control Tower based automatic Dome9 onboarding of the account.
In the document below one can find an explicit design and flow descriptions.
`https://docs.google.com/document/d/1EqOwy2PnU8HxOUy2qDJ9M_wwX35zYFI9FoqAvn7LF6g`

CR entry-points:
1) `template.yaml` is the entry-point used to generate Control Tower Master side infrastructure.
2) `lambda_function.py -> lambda_handler` is the entry point of a triggered lambda.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -11,12 +11,15 @@ Source code for AWS serverless application to automatically onboard AWS account
...
 
-## Installation
+## Installation with AWS Console
 1. Generate CloudGuard Dome9 API keys. You can find the instructions [here](https://supportcenter.checkpoint.com/supportcenter/portal?eventSubmit_doGoviewsolutiondetails=&solutionid=sk144514&partition=General&product=CloudGuard)
-2. User this to launch the application stack installation: <br/> 
-[![Launch Stack](https://s3.amazonaws.com/cloudformation-examples/cloudformation-launch-stack.png)](https://console.aws.amazon.com/cloudformation/home#/stacks/create/review?stackName=dome9-automatic-onboarding&templateURL=https://test-control-tower-integration.s3.amazonaws.com/template.yaml)
+2. Login to your AWS Console
+3. Navigate to Lambda -> Create Function
+4. Select "Browse serverless app repository option"
+5. Type "dome9" and the search text box and make sure that "Show apps that create custom IAM roles or resource policies" is selected
```
Comment (erez-dome9) [inline README.md:21]: rephrase

File: lambda_function.py [hunk 2]
```diff
@@ -1,9 +1,161 @@
...
+from uuid import uuid4
+from typing import Dict
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+
+class Onboarder(object):
+
+    MASTER_ACCOUNT_PERMISSIONS_STACK_SET_NAME = 'Dome9AutomaticOnboardigStackSet'
```
Comment (erez-dome9) [inline lambda_function.py:17]: typo: Onboardig

File: lambda_function.py [hunk 3]
```diff
@@ -1,9 +1,161 @@
...
+from dome9_type_annotations.client import Client
+from resources.aws_cloud_account import CloudAccount, CloudAccountCredentials
+from uuid import uuid4
+from typing import Dict
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+
+class Onboarder(object):
```
Comment (erez-dome9) [inline lambda_function.py:15]: rename Class name

File: lambda_function.py [hunk 4]
```diff
@@ -1,9 +1,161 @@
...
+
+    MASTER_ACCOUNT_PERMISSIONS_STACK_SET_NAME = 'Dome9AutomaticOnboardigStackSet'
+    CUSTOMER_ACCOUNT_EXECUTION_ROLE_NAME = "AWSControlTowerExecution"
+    MASTER_ACCOUNT_STACK_SET_ROLE = "service-role/AWSControlTowerStackSetRole"
+    STACK_OPERATION_WAIT_RETRIES = 12
+    STACK_OPERATION_WAIT_SLEEP = 10
+
+    def __init__(self, region_name: str, customer_account_id: str, customer_account_name: str, readonly: bool = True) -> None:
+        logger.info(f"Initing onboarder with region_name: '{region_name}', customer_account_id: '{customer_account_id}', customer_account_name: '{customer_account_name}'")
```
Comment (erez-dome9) [inline lambda_function.py:24]: typo: Initing

Comment (erez-dome9) [inline lambda_function.py:24]: we use single quote. in such case we do:
`logger.info(f'''my name is: '{name}' and i rock''')`

File: lambda_function.py [hunk 5]
```diff
@@ -1,9 +1,161 @@
...
-	print("Received event: " + json.dumps(event, indent=2))
-	return event  # Echo back the first key value
-	#raise Exception('Something went wrong')
+    logger.info(f"Event: {event}")
+    logger.info(f"Context: {context}")
+
+    event = event["detail"]
+
+    logger.info(f"Event reported state: {event['serviceEventDetails']['createManagedAccountStatus']['state']}'")
+    onboarder = Onboarder(event["awsRegion"], event["serviceEventDetails"]["createManagedAccountStatus"]["account"]["accountId"], event["serviceEventDetails"]["createManagedAccountStatus"]["account"]["accountName"])
```
Comment (erez-dome9) [inline lambda_function.py:154]: add the positional args.
example:
`Onboarder(arg1='value1', arg2='value2')`

File: lambda_function.py [hunk 6]
```diff
@@ -1,9 +1,161 @@
...
+        register_result = self.register_to_dome9()
+        return register_result
 
 
 def lambda_handler(event, context):
-	print("Received event: " + json.dumps(event, indent=2))
-	return event  # Echo back the first key value
-	#raise Exception('Something went wrong')
+    logger.info(f"Event: {event}")
+    logger.info(f"Context: {context}")
```
Comment (erez-dome9) [inline lambda_function.py:149]: no need anymore

File: lambda_function.py (hunk 5 above)
Comment (erez-dome9) [inline lambda_function.py:154]: extract the dict values to separate parameters.
`account_id = event["serviceEventDetails"]["createManagedAccountStatus"]["account"]["accountId"]`

File: lambda_function.py [hunk 7]
```diff
@@ -1,9 +1,161 @@
...
+        self.customer_account_name = customer_account_name
+        self.customer_account_new_role_name = f"Dome9Role-{self.master_account_id}-{self.customer_account_id}"
+        self.customer_account_new_role_arn = f"arn:aws:iam::{self.customer_account_id}:role/{self.customer_account_new_role_name}"
+        self.customer_account_external_id = self.generate_external_id()
+
+        self.cloudformation_client = boto3.client("cloudformation")
+        if readonly:
+            self.user_side_stack_cf_filename = "user_side_stack.yaml"
+        else:
+            self.user_side_stack_cf_filename = "user_side_stack.yaml"
```
Comment (erez-dome9) [inline lambda_function.py:38]: same file name in both cases

File: lambda_function.py [hunk 8]
```diff
@@ -1,9 +1,161 @@
...
+
+        response = self.cloudformation_client.create_stack_set(
+            StackSetName=Onboarder.MASTER_ACCOUNT_PERMISSIONS_STACK_SET_NAME,
+            Description='Dome9 auto onboarding stack set',
+            TemplateBody=template_body,
+            Parameters=[{'ParameterKey': 'Externalid', 'ParameterValue': "Placeholder"},
+                        {'ParameterKey': 'AccountRoleName', 'ParameterValue': "Placeholder"}],
+            Capabilities=[
+                'CAPABILITY_IAM', 'CAPABILITY_NAMED_IAM', 'CAPABILITY_AUTO_EXPAND'
+            ],
```
Comment (erez-dome9) [inline lambda_function.py:69]: you should run our `yapf` to align the code styling

File: lambda_function.py [hunk 9]
```diff
@@ -1,9 +1,161 @@
...
+            ParameterOverrides=[{'ParameterKey': 'AccountRoleName', 'ParameterValue': self.customer_account_new_role_name},
+                                {'ParameterKey': 'Externalid', 'ParameterValue': self.customer_account_external_id}],
+            OperationPreferences={
+                'FailureToleranceCount': 0,
+                'MaxConcurrentCount': 3,
+            })
+
+        self.wait_for_stack_operation(response["OperationId"])
+
+    def wait_for_stack_operation(self, operation_id: str) -> None:
```
Comment (erez-dome9) [inline lambda_function.py:94]: shouldn't raise something if the wait operation exceeds the retries?
add a raise exception at the end of the function (outside the for loop)

File: lambda_function.py [hunk 10]
```diff
@@ -1,9 +1,161 @@
...
+
+    def delete_stack_set(self):
+        logger.info(f"Deleting StackSet: {self.MASTER_ACCOUNT_PERMISSIONS_STACK_SET_NAME}")
+        return self.cloudformation_client.delete_stack_set(StackSetName=self.MASTER_ACCOUNT_PERMISSIONS_STACK_SET_NAME)
+
+    def register_to_dome9(self) -> Dict:
+        dome9_client = Client()
+        credentials = CloudAccountCredentials(arn=self.customer_account_new_role_arn, secret=self.customer_account_external_id)
+        payload = CloudAccount(name=self.customer_account_name, credentials=credentials)
+        resp = dome9_client.aws_cloud_account.create(body=payload)
```
Comment (erez-dome9) [inline lambda_function.py:122]: resp / response. make it consistent all across

File: lambda_function.py [hunk 11]
```diff
@@ -1,9 +1,161 @@
...
+        try:
+            self.create_stack_set()
+        except Exception as e:
+            if "NameAlreadyExistsException" in repr(e):
+                logger.info(f"Received NameAlreadyExistsException for stack_set '{self.MASTER_ACCOUNT_PERMISSIONS_STACK_SET_NAME}'. Error: {repr(e)}")
+                self.delete_stack_instances()
+                time.sleep(20)
+                #self.delete_stack_set()
+                #time.sleep(30)
+                #self.create_stack_set()
```
Comment (erez-dome9) [inline lambda_function.py:135]: remove

File: lambda_function.py [hunk 12]
```diff
@@ -1,9 +1,161 @@
...
+        payload = CloudAccount(name=self.customer_account_name, credentials=credentials)
+        resp = dome9_client.aws_cloud_account.create(body=payload)
+        return resp
+
+    def create_stack_set_flow(self) -> None:
+        try:
+            self.create_stack_set()
+        except Exception as e:
+            if "NameAlreadyExistsException" in repr(e):
+                logger.info(f"Received NameAlreadyExistsException for stack_set '{self.MASTER_ACCOUNT_PERMISSIONS_STACK_SET_NAME}'. Error: {repr(e)}")
```
Comment (erez-dome9) [inline lambda_function.py:130]: isn't it `exception` level?

File: lambda_function.py [hunk 13]
```diff
@@ -1,9 +1,161 @@
...
+                self.delete_stack_instances()
+                time.sleep(20)
```
Comment (erez-dome9) [inline lambda_function.py:132]: assume that u are working on this sleep...

File: lambda_function.py (hunk 8 above)
Comment (AlexeyBeley) [reply lambda_function.py]: done

File: lambda_function.py (hunk 9 above)
Comment (AlexeyBeley) [reply lambda_function.py:117]: Yes, all the `waiting` will be redesigned in the next PR.

File: lambda_function.py (hunk 12 above)
Comment (AlexeyBeley) [reply lambda_function.py:185]: I though to do it. But it's ignored now - both exceptions got to regular stdout without formatting.

File: lambda_function.py (hunk 13 above)
Comment (AlexeyBeley) [reply lambda_function.py:187]: You are right. Will be fixed with waiter

File: lambda_function.py (hunk 11 above)
Comment (AlexeyBeley) [reply lambda_function.py]: done

File: lambda_function.py (hunk 6 above)
Comment (AlexeyBeley) [reply lambda_function.py]: done

File: lambda_function.py (hunk 5 above)
Comment (AlexeyBeley) [reply lambda_function.py]: done

Comment (AlexeyBeley) [reply lambda_function.py]: As discussed.

**Pull request merged**